### PR TITLE
perf: Port rpc changes from corefx

### PR DESCRIPTION
**Pull request by Wraith2**
This is a port of https://github.com/dotnet/corefx/pull/34049 which changes how rpc object are constructed and re-used so avoid repeated creation of new SqlParameter instances where they aren't needed.

This change was not brought across from corefx initially because the introduction of AlwaysEncrypted required additionally changes. Those changes are part of this new PR. manual tests pass in debug mode (with all asserts) so it looks like the AE changes are ok but I'd like some oversight to make sure I've understood what it's doing correctly. 

I think I've found a weakness in the batch rpc mode with AE. It assumes that the command will be in sysemParams[0] but this is only true if the batch rpc was constructed as an sql command and not as a stored procedure. This isn't currently causing a problem but might need tightening up.

### CODE REVIEW

**Comment by cheenamalhotra** [comment]
Hi @Wraith2 

Please check and confirm code coverage (as close to 100% as possible) on changes made here. If RPC tests don't exist, please add them to the suite so they cover maximum line changes here.

**Comment by Wraith2** [comment]
How do I check coverage? Given that this effects how you do queries it'd be quite surprising if it wasn't covered.

**Comment by Wraith2** [comment]
I'm definitely going to need help getting coverage. I've tried getting the manual tests running in visual studio and i'm unable to apply the required environment variables (launchsettings.json doesn't work for some reason) and the tests themselves won't execute because of an nuint serialization exception. So either I need some guidance on how you get it working in visual studio or how to get coverage numbers from msbuild and use them sensibly.

**Comment by cheenamalhotra** [comment]
Sure, I'd expect it to be covered too, we just want to verify if we have everything covered here. So maybe try running tests in Visual Studio with coverage plugins?

Also extracted this coverage report for this PR, hope it serves well as a starting point.
[20190922.2...13627.coverage.zip](https://github.com/dotnet/SqlClient/files/3654346/20190922.2.13627.coverage.zip)

**Comment by cheenamalhotra** [comment]
FYI, you can also run `dotnet test` command with appended `--collect:"code coverage" --collect-format:"coverage" ` or `--collect:"code coverage" --collect-format:"coveragexml"` that will give you similar file for your specific test run.

**Comment by Wraith2** [comment]
Should tests run from within visual studio? When I try I get this error which stops the entire run:

```
[25/09/2019 11:13:44.175 PM Informational] Store opened in 0.320 sec.
[25/09/2019 11:13:45.071 PM Informational] ---------- Discovery started ----------
[25/09/2019 11:13:45.285 PM Error] Microsoft.VisualStudio.TestPlatform.ObjectModel.TestPlatformException: Unable to find E:\Programming\csharp7\SqlClient\bin\Windows_NT\Debug.AnyCPU\CoreFx.Private.TestUtilities\netcoreapp2.1\testhost.dll. Please publish your test project and retry.
   at Microsoft.VisualStudio.TestPlatform.CrossPlatEngine.Hosting.DotnetTestHostManager.GetTestHostPath(String runtimeConfigDevPath, String depsFilePath, String sourceDirectory)
   at Microsoft.VisualStudio.TestPlatform.CrossPlatEngine.Hosting.DotnetTestHostManager.GetTestHostProcessStartInfo(IEnumerable`1 sources, IDictionary`2 environmentVariables, TestRunnerConnectionInfo connectionInfo)
   at Microsoft.VisualStudio.TestPlatform.CrossPlatEngine.Client.ProxyOperationManager.SetupChannel(IEnumerable`1 sources)
   at Microsoft.VisualStudio.TestPlatform.CrossPlatEngine.Client.ProxyDiscoveryManager.DiscoverTests(DiscoveryCriteria discoveryCriteria, ITestDiscoveryEventsHandler2 eventHandler)
[25/09/2019 11:13:45.899 PM Informational] [xUnit.net 00:00:00.3081508]   Discovering: Microsoft.Data.SqlClient.Tests
[25/09/2019 11:13:45.970 PM Informational] [xUnit.net 00:00:00.3776758]   Discovering: Microsoft.Data.SqlClient.ManualTesting.Tests
[25/09/2019 11:13:46.350 PM Informational] [xUnit.net 00:00:00.6369007]   Discovered:  Microsoft.Data.SqlClient.ManualTesting.Tests
[25/09/2019 11:13:46.456 PM Informational] [xUnit.net 00:00:00.7167884]   Discovered:  Microsoft.Data.SqlClient.Tests
[25/09/2019 11:13:46.962 PM Warning] No test is available in E:\Programming\csharp7\SqlClient\bin\Windows_NT\Debug.AnyCPU\Microsoft.Data.SqlClient\netfx\Microsoft.Data.SqlClient.dll. Make sure that test discoverer & executors are registered and platform & framework version settings are appropriate and try again.
[25/09/2019 11:13:47.024 PM Informational] ========== Discovery aborted: -1 tests found (0:00:01.9150349) ==========
```

I can't use the coverage file you provided because I'm only on visual studio pro. Code coverage is an enterprise or ultimate feature and since I buy my own visual studio license rather than getting it through an employer or volume license both editions are way out of my reach. If I could get the test running resharper should provide coverage, but that hits the error above.

**Comment by cheenamalhotra** [comment]
Hi @Wraith2 
Try running in Release configuration, we're aware of Debug Configuration test issues in Visual Studio.

**Comment by Wraith2** [comment]
None of the release mode configurations work for me either. I get an exception from inside xunit extensions. The coverage cli commands suggested aren't recognised at the commandline when appended to the test commands and the dotnet cli docs don't list them even though I can find the feature request for them being added, perhaps post 2.1 dotnet only. using /p:CollectCoverage=true doesn't seem to do anything either.

I've opened a couple of tracking issues for test and coverage. At the moment I can't provide any coverage data.

**Comment by Wraith2** [comment]
Now that 1.1.0 is out can this be reviewed please? It's been rebased and is only failing on that AE tests that fail in master as well. This work is needed for the batch implementation.

**Comment by cheenamalhotra** [comment]
Hi @Wraith2 

Yes, please update your branch with changes from master to trigger a fresh CI build.

**Comment by Wraith2** [comment]
Updated.

**Comment by cheenamalhotra** [comment]
Thanks @Wraith2 

It looks like the Enclave Provider tests are failing for .NET Core, please check logs in the checks (they're public now) and review implementation to find related code changes. It's tricky to get the setup for Enclave Provider for debugging, so I'd recommend going down your changes and understanding failing tests to fix what went wrong.

You'd be the best person to review first and fix the issues being creator of the PR, we'll start reviewing once we have all tests passing and changes are ready to check-in. If you need any help let us know.

**Comment by Wraith2** [comment]
So does SQL17W-E mean with enclave? I can guess from stack traces but realistically if I'm (any anyone else) going to need to ensure enclaves work we'll need instructions on how to set up and run those tests correctly, can you provide something?

**Comment by cheenamalhotra** [comment]
Sure, you can begin from here: https://docs.microsoft.com/en-us/sql/relational-databases/security/tutorial-getting-started-with-always-encrypted-enclaves?view=sql-server-ver15 for setup, and use following tutorials to setup .NET Applications, same instructions apply for Microsoft.Data.SqlClient, except that Enclave provider is internally available and does not need additional package reference.

We are also just-in-process of putting specific documentation for Microsoft.Data.SqlClient, since now Enclave provider is available in GA 1.1

**Comment by Wraith2** [comment]
The changes are done and pass locally on the same tests as master. The SQL12E-E leg is failing with timeouts and flaky tests from what I can see, can you retry that leg?

**Comment by cheenamalhotra** [comment]
Hi @Wraith2 

Please trigger a new build by pushing a commit. DevOps recently has some issues with Rebuild.

**Comment by Wraith2** [comment]
New build triggered and looks like it's been stuck overnight.

File: src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SqlCommand.cs [hunk 1]
```diff
@@ -5253,21 +5247,24 @@ private int CountSendableParameters(SqlParameterCollection parameters)
         // Returns total number of parameters
         private int GetParameterCount(SqlParameterCollection parameters)
         {
-            return ((null != parameters) ? parameters.Count : 0);
+            return (null != parameters) ? parameters.Count : 0;
         }
 
+
```
[the remark this answers is not in the document]
Comment (Wraith2) [reply src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SqlCommand.cs]: Fixed.

File: src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SqlCommand.cs [hunk 2]
```diff
@@ -5140,20 +5122,23 @@ private void SetUpRPCParameters(_SqlRPC rpc, int startCount, bool inSchema, SqlP
...
                 if (ShouldSendParameter(parameter))
                 {
-                    rpc.parameters[j] = parameter;
+                    byte options = 0;
 
                     // set output bit
-                    if (parameter.Direction == ParameterDirection.InputOutput ||
-                        parameter.Direction == ParameterDirection.Output)
-                        rpc.paramoptions[j] = TdsEnums.RPC_PARAM_BYREF;
+                    if (
```
[the remark this answers is not in the document]
Comment (Wraith2) [reply src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SqlCommand.cs]: Made single line.

File: src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SqlCommand.cs [hunk 3]
```diff
@@ -3809,7 +3815,9 @@ private void ReadDescribeEncryptionParameterResults(SqlDataReader ds, ReadOnlyDi
 
                                     // This is effective only for BatchRPCMode even though we set it for non-BatchRPCMode also,
                                     // since for non-BatchRPCMode mode, paramoptions gets thrown away and reconstructed in BuildExecuteSql.
-                                    rpc.paramoptions[paramIdx] |= TdsEnums.RPC_PARAM_ENCRYPTED;
+                                    int options = (int)(rpc.userParamMap[index] >> 32);
+                                    options |= TdsEnums.RPC_PARAM_ENCRYPTED;
+                                    rpc.userParamMap[index] =  ((((long)options) << 32) | (long)index);
```
[the remark this answers is not in the document]
Comment (Wraith2) [reply src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SqlCommand.cs]: Fixed.

File: src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SqlCommand.cs [hunk 4]
```diff
@@ -5394,46 +5393,48 @@ private SqlParameter BuildStoredProcedureStatementForColumnEncryption(string sto
...
-                if (parameters[i].Direction != ParameterDirection.ReturnValue)
+                if (parameters[index].Direction != ParameterDirection.ReturnValue)
                 {
-                    execStatement.AppendFormat(@", {0}={0}", parameters[i].ParameterNameFixed);
+                    execStatement.AppendFormat(@", {0}={0}", parameters[index].ParameterNameFixed);
 
                     // InputOutput and Output parameters need to be marked as such.
-                    if (parameters[i].Direction == ParameterDirection.Output ||
-                    parameters[i].Direction == ParameterDirection.InputOutput)
+                    if (
```
[the remark this answers is not in the document]
Comment (Wraith2) [reply src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SqlCommand.cs]: made single line.

**Pull request merged**